inc/event_listeners.h

```diff
@@ -0,0 +1,48 @@
+#ifndef EVENT_LISTENERS_H
+#define EVENT_LISTENERS_H
+
+#include <vector>
+#include <string>
+#include <iostream>
+#include <tuple>
+
+enum Event {
+  BEGIN_PHASE,
+  DO_PREDICT_BRANCH
+};
+
+#include "listeners/cpi_printout.h"
+#include "listeners/test_listener.h"
+
+inline std::vector<bool> listener_activation_map;
+inline auto listeners = std::make_tuple(new CPIPrintout(), new TestListener());
+inline std::vector<std::string> listener_names = {"CPIPrintout", "TestListener"};
+
+inline void init_event_listeners(const std::vector<std::string>& requested_listeners) {
+  listener_activation_map = std::vector<bool>(listener_names.size(), 0);
+  for (std::string name : requested_listeners) {
+    bool found = false;
+    for (size_t i = 0; i < listener_names.size(); i++) {
+      if (listener_names[i] == name) {
+        listener_activation_map[i] = true;
+        found = true;
+        break;
+      }
+    }
+    if (!found) {
+      std::cout << "WARNING: Listener \"" << name << "\" not found\n";
+    }
+  }
+}
+
+template<Event e, typename... Args>
+void handle_event(Args... args) {
+  if (listener_activation_map[0]) {
+    std::get<0>(listeners)->handle_event<e>(args...);
+  }
+  if (listener_activation_map[1]) {
+    std::get<1>(listeners)->handle_event<e>(args...);
+  }
+}

```

Comment:
Are you ready for this? Tuples are fun sometimes.
```suggestion
template<Event e, std::size_t Idx, typename... Args>
void handle_listener_event(Args... args) {
    if (listener_activation_map[Idx]) {
        std::get<Idx>(listeners)->template handle_event<e>(args...);
    }
}

template<Event e, typename... Args>
void handle_event(Args... args) {
    [&] <std::size_t... Is> (std::index_sequence<Is...>){
        (handle_listener_event<e, Is>(args...), ...);
    }(std::make_index_sequence<std::tuple_size_v<decltype(listeners)>>{}); // immediately invoked
}
```

Reply:
I think it should be illegal to combine lambdas, templates, and variadic arguments all at the same time. The effect is very convenient, though. I didn't realize variadic arguments could extend to calling functions.